clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala

```diff
@@ -42,16 +42,27 @@ class Exporter(spark : SparkSession, apiClient: ApiClient, filter: KeyFilter, re
     val hadoopConf = spark.sparkContext.hadoopConfiguration
     val serializedConf = new SerializableWritable(hadoopConf)
     val f = filter
-    val errs = actionsDF.map(
-      row =>
-        Exporter.handleRow(f, round, ns, rel, serializedConf, row)
+    var par = parallelism
+    val errs = actionsDF.mapPartitions( part => {
+      val pool: ExecutorService = Executors.newFixedThreadPool(par)
+
+      val callableTasks = new util.ArrayList[Callable[ExportStatus]]
+      while (part.hasNext) {
+        val row = part.next()
+        callableTasks.add(new Handler(f, round, ns, rel, serializedConf, row))
+      }
+
+      val res = pool.invokeAll(callableTasks).asScala.map(f => f.get()).iterator
```

Comment:
Why so procedural?  I think this should be enough (you may need to get an iterator from `res`, but that's about all).
```suggestion
      val res = pool.invokeAll(part.map(row => callableTasks.add(new Handler(f, round, ns, rel, serializedConf, row))))
```

![Joker asks: Why so procedural?](https://user-images.githubusercontent.com/7802932/113701445-d7a9c400-96e0-11eb-92d3-bc76fc806070.png)

Reply:
HAHA Done.